open_instruct/vllm_utils3.py

```diff
@@ -190,55 +205,128 @@ def __init__(
+                tokens_prompt = vllm.TokensPrompt(prompt_token_ids=prompt)
+                self.llm_engine.add_request(request_id, tokens_prompt, request.sampling_params)
+
+            # Run the engine event loop until all requests are finished
+            outputs = []
+            while self.llm_engine.has_unfinished_requests():
+                step_outputs = self.llm_engine.step()
+                for output in step_outputs:
+                    if output.finished:
+                        outputs.append(output)
+
+            # Sort outputs by prompt index (i)
+            # Request IDs are now in format: batch_{step}_{i}
+            outputs.sort(key=lambda x: int(x.request_id.split("_")[-1]))
+
+            # Debug logging
+            total_responses = sum(len(output.outputs) for output in outputs)
+            self.logger.info(
+                f"[LLMRayActor] Collected {len(outputs)} RequestOutputs with "
+                f"{total_responses} total responses (expected {len(prompts) * n_samples})"
+            )
 
-    def process_from_queue(self, num_training_steps=None, resume_training_step=1, timeout=0.1):
+            # Additional debug info
+            for i, output in enumerate(outputs):
+                self.logger.info(
+                    f"[LLMRayActor] RequestOutput {i} (request_id={output.request_id}): "
+                    f"has {len(output.outputs)} completions"
+                )
+
+            result = self._process_outputs(outputs, dataset_index=request.dataset_index)
+            try:
+                if request.is_eval:
+                    self.eval_results_queue.put(result, timeout=10)
+                else:
+                    self.results_queue.put(result, timeout=10)
+            except queue.Full:
+                self.logger.warning("Results queue is full, discarding result.")
+
+    def process_from_queue(self, timeout=0.1):
         """Process a single element from the queue."""
-        try:
-            request = self.prompt_queue.get(timeout=timeout)
-            # Use generation_config from the request
-            result = self._generate_batch(
-                request.prompts,
-                request.generation_config,
-                dataset_index=request.dataset_index,
-                training_step=request.training_step,
-            )
-            if request.is_eval:
-                self.eval_results_queue.put(result, timeout=1)
-            else:
-                self.results_queue.put(result, timeout=1)
-            return 1  # Successfully processed a request
-        except queue.Empty:
-            self.logger.warning("[LLMRayActor] No request in the queue to process. Returning from process_from_queue.")
-            return 0  # No request to process
-        except queue.Full:
-            self.logger.warning(f"[LLMRayActor] Results queue is full. Skipping insert. {request.is_eval=}.")
-            return 0
-
-    def _generate_batch(
+        if self.tool_use:
```

Comment:
I feel like a nicer way to do this might just to have the toolvllm subclass / be an alternate class to llmrayactor and override the generation loop?